TFNetworkRecLayer.py

```diff
@@ -5955,3 +5955,72 @@ def _get_output_native_rec_op(self, cell):
       y = y[0]
 
     return y
+
+
+class ZoneoutLSTMCell(BaseRNNCell):
+  '''Wrapper for tf LSTM to create Zoneout LSTM Cell
+
+  inspired by:
+  https://github.com/teganmaharaj/zoneout/blob/master/zoneout_tensorflow.py
+
+  Published by one of 'https://arxiv.org/pdf/1606.01305.pdf' paper writers.
+
+  Many thanks to @Ondal90 for pointing this out. You sir are a hero!
+  '''
+
+  def __init__(self, num_units, zoneout_factor_cell=0., zoneout_factor_output=0., state_is_tuple=True, name=None, **kwargs):
+    '''Initializer with possibility to set different zoneout values for cell/hidden states.
+    '''
+    zm = min(zoneout_factor_output, zoneout_factor_cell)
+    zs = max(zoneout_factor_output, zoneout_factor_cell)
+
+    if zm < 0. or zs > 1.:
+      raise ValueError('One/both provided Zoneout factors are not in [0, 1]')
+
+    self._cell = tf.nn.rnn_cell.LSTMCell(num_units, state_is_tuple=state_is_tuple, name=name)
+    self._zoneout_cell = zoneout_factor_cell
+    self._zoneout_outputs = zoneout_factor_output
+    if kwargs.get('is_training', None) is None:
+      from TFNetwork import TFNetwork
+      self.is_training = TFNetwork.get_current_network().train_flag
+    else:
+      self.is_training = kwargs['is_training']
+
+    self.state_is_tuple = state_is_tuple
+
+  @property
+  def state_size(self):
+    return self._cell.state_size
+
+  @property
+  def output_size(self):
+    return self._cell.output_size
+
+  def __call__(self, inputs, state, scope=None):
+    '''Runs vanilla LSTM Cell and applies zoneout.
+    '''
+    # Apply vanilla LSTM
+    output, new_state = self._cell(inputs, state, scope)
+
+    if self.state_is_tuple:
+      (prev_c, prev_h) = state
+      (new_c, new_h) = new_state
+    else:
+      num_proj = self._cell._num_units if self._cell._num_proj is None else self._cell._num_proj
+      prev_c = tf.slice(state, [0, 0], [-1, self._cell._num_units])
+      prev_h = tf.slice(state, [0, self._cell._num_units], [-1, num_proj])
+      new_c = tf.slice(new_state, [0, 0], [-1, self._cell._num_units])
+      new_h = tf.slice(new_state, [0, self._cell._num_units], [-1, num_proj])
```

Comment:
Is there any case where `not state_is_tuple` is better? If not, I think we should just remove that.